tutorials/haskell/gui-application/tutorial.md

```diff
@@ -0,0 +1,825 @@
+  in Value x $ Just $
+    case action of
+      Nothing -> op ""
+      Just  a -> op (getSndArg a)
+
+-- | Change state as if current argument is removed.
+clearEntry :: Value -> Value
+clearEntry (Value x action) =
+  case action of
+    Nothing -> Value "" Nothing
+    Just  a ->
+      if null (getSndArg a)
+        then Value "" Nothing
+        else Value x (Just $ mapAction (const "") a)
+
+-- | Change state returning it to the default value.
+clearAll :: Value -> Value
+clearAll = const (Value "" Nothing)
+
+-- | Evaluate current calculator's state putting result in place of first
+-- argument.
+equals :: Value -> Value
+equals (Value x action) =
+  case action of
+    Nothing -> Value x Nothing
+    Just  a ->
+      if null (getSndArg a)
+        then Value x action
+        else Value result Nothing
+          where
+            g  :: String -> Double
+            g ""       = 0
+            g ('.':xs) = g ('0':'.':xs)
+            g xs       = read (reverse xs)
+            x' = g x
+            y' = g (getSndArg a)
+            result = reverse . show $
+              case a of
+                Addition       _ -> x' + y'
+                Subtraction    _ -> x' - y'
+                Multiplication _ -> x' * y'
+                Division       _ -> x' / y'
+```
+
+The calculator is not perfect, but good enough for our purposes. Compile,
+run it and see how it works. Implementation of the rest of functionality is
+left as an exercise for the reader.
+
+## Using Glade to design forms
+
+You probably have noticed that manual creation of forms introduces quite a
+bit of boilerplate in our code. This can become even worse as your forms get
+more complex. Because of this, I think it's time to try our hand on a modern
+UI designer for GTK+ called [Glade](https://glade.gnome.org/).
+
+Glade is straightforward to install and use. Open the application, you will
+see panels with various widgets: top-level objects (such as window),
+containers, and controls (such as buttons).
+
+Short plan how to re-create our calculator form with Glade:
+
+1. Select window button on the topmost palette and the window will appear on
+   the working area.
+
+2. Enter “Calculator” in the title field.
+
+3. Don't forget to fill out the “ID” attribute of every widget, this is how
+   you will access widgets on your form in the Haskell code.
+
+4. Create grid with id `grid`. When asked about number of rows and columns,
+   choose 7 × 5. Select the “Homogeneous” check box under the “Rows” title.
+
+5. Insert entry, use “Drag and resize widgets in the workspace” button on
+   the top tool bar to make it occupy the entire top row.
+
+6. Insert buttons to match our existing design.
+
+![Building the calculator form in Glade](/tutorials/haskell/gui-application/calc-3-mini.png)
+
+(Here is a [bigger image](/tutorials/haskell/gui-application/calc-3.png).)
+
+*Hint: complete glade form of our calculator is available under the Stack
+Builders tutorial repository.*
```

Comment:
Perhaps you didn't include the link on purpose, but we should make sure the link to the code is somewhere other than the tutorial list.